Feature/bruker data

jspaezp opened the pull request:
<!--
# nf-core/quantms pull request

Many thanks for contributing to nf-core/quantms!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/quantms/tree/master/.github/CONTRIBUTING.md)
-->

## TODO

- [x] Add support for non-archived .d files
- [ ] Docs
- [x] inline docs
- [x] Make input naming ocnsistent (right now a lot of places are called mzml and shoudl be ms_file)

## PR checklist

- [x] This comment contains a description of changes (with reason).
    - This PR adds support for .d 'files' and passing them directly to DIA-NN for DIA search. As well as related changes to support intermediate files not being exclusively mzMLs
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/quantms/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, also make a PR on the nf-core/quantms _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`). **Note** locally this does not run anything.
- [x] Usage Documentation in `docs/usage.md` is updated. Depends on https://github.com/nf-core/quantms/pull/103
- [x] Output Documentation in `docs/output.md` is updated. Depends on https://github.com/nf-core/quantms/pull/103
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).
    - Should a 'Contribtors' section be added and list myself in it?

ypriverol commented:
@jspaezp this PR from nf-core templates https://github.com/nf-core/quantms/pull/103 would be good to merged into your branch. While it creates some unnecessary work, it is better to integrate in an early stage than later to avoid the need to change more files when we go for release.

jspaezp commented on bin/diann_convert.py:
```diff
@@ -23,13 +36,14 @@ def cli():
 @click.option("--qvalue_threshold", "-q", type=float)
 @click.pass_context
-def convert(ctx, folder, dia_params, diann_version, charge, missed_cleavages, qvalue_threshold):
+def convert(ctx, folder, exp_design, dia_params, diann_version, charge, missed_cleavages, qvalue_threshold):
```
@ypriverol the code convention in this repo is 120 columns, right?

ypriverol replied:
Forget some of the messages. We added sonatype some time ago but we are not paying to much attetion to it. We try to use more the PythonBlack which is required by nf-core.

jspaezp commented:
@ypriverol Thanks for the heads up, I will work on adding it

jspaezp commented on bin/diann_convert.py:
```diff
@@ -910,14 +1139,227 @@ def calculate_mz(seq, charge):
+            cov = calculate_coverage(fasta_id_to_seqs[f_id], ids_to_seqs[acc_to_ids[acc]])
+            out_cov = format(cov, ".03f")
+
+        out[i] = out_cov
```
@sonatype-lift exclude issue

jspaezp commented:
@sonatype-lift exclude E501

jspaezp commented on bin/diann_convert.py:
This file has very aggressively optimized, since it was the slowest part of the whole pipeline for some of our data.
In addition, a lot of logging has been added to debug failures and track run performance.

ypriverol replied:
Thanks a lot for your contribution!!! I will review the changes to that particular file.

jspaezp replied:
This change is large enough that it might deserve a PR of its own.

jspaezp commented on bin/diann_convert.py:
```diff
@@ -440,46 +597,71 @@ def mztab_PRH(report, pg, index_ref, database, fasta_df):
     protein_details_df = (
         protein_details_df.drop("accession", axis=1).join(prh_series).reset_index().drop(columns="index")
     )
+    # Q: how is the next line different from `df.loc[:, "col"] = 'protein_details'` ??
```
Question

WangHong007 replied:
No difference here.

jspaezp replied:
Thanks!

jspaezp commented on bin/diann_convert.py:
```diff
@@ -440,46 +597,71 @@ def mztab_PRH(report, pg, index_ref, database, fasta_df):
+    # out_mztab_PRH has columns accession and Protein.Ids; 'Q9NZJ9', 'A0A024RBG1;Q9NZJ9;Q9NZJ9-2']
+    # the report table has 'Protein.Group' and 'Protein.Ids': 'Q9NZJ9', 'A0A024RBG1;Q9NZJ9;Q9NZJ9-2'
+    # Oddly enough the last implementation mapped the the accession (Q9NZJ9) in the mztab
+    # to the Protein.Ids (A0A024RBG1;Q9NZJ9;Q9NZJ9-2), leading to A LOT of missing values.
```
I am also not sure if this was the intended behavior

jspaezp commented on bin/diann_convert.py:
```diff
@@ -910,14 +1132,227 @@ def calculate_mz(seq, charge):
     :type charge: int
     :return:
     """
+    # Q: is this faster if we make it a set? and maybe make it a global variable?
```
This question is not relevant anymore, since the function is no longer in use.

jspaezp commented on bin/diann_convert.py:
```diff
@@ -910,14 +1132,227 @@ def calculate_mz(seq, charge):
+    In my benchmarking it went from 35.23 seconds for 4 samples, 4 conditions to
+    0.007 seconds.
+
+    This implementation differs in several aspects in the output values:
```
LMK if this is a viable change

jspaezp commented on modules/local/diann_preliminary_analysis/main.nf:
```diff
@@ -7,7 +7,7 @@ process DIANN_PRELIMINARY_ANALYSIS {
         'biocontainers/diann:v1.8.1_cv1' }"
 
     input:
```
TODO: rename to ms_files, since not only mzmls are passed

jspaezp commented on modules/local/diann_preliminary_analysis/main.nf:
```diff
@@ -7,7 +7,7 @@ process DIANN_PRELIMINARY_ANALYSIS {
 
     input:
-    tuple val(meta), file(mzML), file(predict_tsv)
+    tuple val(meta), path(mzML), path(predict_tsv)
```
also rename the predict tsv, since it is a .speclib binary. If I am not wrong

jspaezp commented on modules/local/diannsummary/main.nf:
```diff
@@ -35,9 +39,16 @@ process DIANNSUMMARY {
     diann   --lib ${empirical_library} \\
             --fasta ${fasta} \\
-            --f ${(mzMLs as List).join(' --f ')} \\
+            --f ${(ms_files as List).join(' --f ')} \\
```
would this break if it is joined by `'\n --f \\'`

jpfeuffer replied:
Good question. Not sure when \n actually gets resolved to a newline.
One could try.

jspaezp replied:
That is a problem for later ... RN it is a problem when running on nextflow tower, because lines have a max length (so not all of the command is captured, but it executes correctly)

jspaezp commented on modules/local/pmultiqc/main.nf:
```diff
@@ -25,6 +25,13 @@ process PMULTIQC {
+
+    # Current hack to attempt matching file stems and not file extensions
+    # sed -i -e "s/((.d.tar)|(.d)|(.mzML)|(.mzml))\\t/\\t/g" 
+    sed -i -e "s/.tar\\t/\\t/g"  results/*openms_design.tsv
```
TODO: check if this is getting used

jspaezp replied:
since it is done here `modules/local/sdrfparsing/main.nf`

jspaezp commented on subworkflows/local/file_preparation.nf:
```diff
@@ -23,6 +24,7 @@ workflow FILE_PREPARATION {
     .branch {
         raw: WorkflowQuantms.hasExtension(it[1], 'raw')
         mzML: WorkflowQuantms.hasExtension(it[1], 'mzML')
+        dotD: WorkflowQuantms.hasExtension(it[1], '.d.tar')
```
TODO add branch here with a plain .d and mix them (and add the exception to TDF2MZML)

ypriverol replied:
@jspaezp wouldn't be more interesting to have `.d` gzip rather than `.tar`. Im asking because I have seen most of the compressed files using gzip instead of tar.

jspaezp replied:
I could implement either, We tested locally and since most of the data inside the file is already compressed, gzip offered very little benefit (<10% compression) whilst dramatically increasing the time it took to generate (4x longer if i recall correctly). So in our use case at least it was not worth having the compression.

I could definitely add something like this to the extraction step:
https://gist.github.com/hightemp/5071909#file-bash-aliases-L32-L60

and correspondingly I would just have the branch be something like 
```
        raw: WorkflowQuantms.hasExtension(it[1], 'raw')
        mzML: WorkflowQuantms.hasExtension(it[1], 'mzML')
        dotD: WorkflowQuantms.hasExtension(it[1], '.d.{tar,tar.gz,tar.bz....}')
```

I will add this to the options (I think we should enforce having the .d though. I would be very hard to track file properties if we attempt to allow "myfile.tar.gz")

Will add this in the next commit

ypriverol replied:
@jspaezp interesting the information about why do you use `.tar`. Would be great to support this approach for those ones using other formats. 

```
 raw: WorkflowQuantms.hasExtension(it[1], 'raw')
 mzML: WorkflowQuantms.hasExtension(it[1], 'mzML')
 dotD: WorkflowQuantms.hasExtension(it[1], '.d.{tar,tar.gz,tar.bz....}')
```

jspaezp replied:
added here: https://github.com/bigbio/quantms/pull/275/commits/85f306058263ee8697c42dbb28eebb5f1b6feacc

jspaezp replied:
@ypriverol 
In a semi-related manner. Do you have any suggestion on public data? If not I think we could upload something.
I was thinking on maybe using https://www.ebi.ac.uk/pride/archive/projects/PXD034128 but its phospho, so it might take A WHILE to search. AND it seems like most people are uploading the data as a single .zip with all the files (and I am not sure if we want to support that or if there is a way to stage the files more efficiently).

LMK what you think

ypriverol commented:
Hi @jspaezp I'm reviewing some of the pending PRs and issues of quantms. Are you waiting for something from our side? Do you need any help to finish this PR?

jspaezp commented:
Hello there @ypriverol !

Trying to summarize the state of affairs here:

Blockers:
1. [ ] https://github.com/nf-core/quantms/pull/103 I added comments here that I think should be fixed so I can merge with the version of that branch that has those fixes.
2. [ ] Right now tests fail because of the file extension requirements (mentioned in the comment of `modules/local/sdrfparsing/main.nf`), related to https://github.com/bigbio/pmultiqc/issues/80, I should be able to implement pmultiqc#80, I recall when I looked at it being a larger change than I was expecting but I can clear a day in my calendar to get it done. In addition https://github.com/bigbio/sdrf-pipelines/issues/145 is related (but in theory only #80 is a blocker). 
3. [ ] As mentioned here https://github.com/bigbio/quantms/pull/275#discussion_r1297217581 I have not added testing data, if you don't have a pride repo as a suggestion I can create one with some random testing data (I would need to ask my company to provide it, but I am pretty sure it would be easy to get some hela lysates).
4. [ ] Linting fails because of this line https://github.com/bigbio/quantms/blob/062e6ba37a50afe3bd712bb7ac2a10fe1e2c4ad7/nextflow_schema.json#L900, since it requires strings to not be empty, BUT right now the pipeline works in a way in which... if it is empty, it evaluates to false, else it uses the path in the string as the pre-computed library. Not sure how to deal with this.
4. [ ] I would like feedback on the remaining comments left on `bin/diann_convert.py`, since I am not sure what the intended correct behavior was in some cases.

(I can/should take care of blocker number 2 (pmultiqc#80), the rest either need discussion/resolution)

While blockers are there:
1. All the code can still be reviewed (that includes another round of self-review on my end).
2. Since the changes on `bin/diann_convert.py` are so large, impactful and isolated from the rest I could split them into a PR of its own.
3. If blocker 4 is resolved, the feature of using a pre-computed library could also be PR by itself.

LMK what you think, I will get started with pmultiqc#80 and we can go though the rest.
Thank you for the attention and nudge!
Sebastian

jspaezp commented on modules/local/assemble_empirical_library/main.nf:
```diff
@@ -22,14 +23,25 @@ process ASSEMBLE_EMPIRICAL_LIBRARY {
     def args = task.ext.args ?: ''
-    mass_acc_ms1 = meta.precursor_mass_tolerance_unit == "ppm" ? meta.precursor_mass_tolerance : 5
-    mass_acc_ms2 = meta.fragment_mass_tolerance_unit == "ppm" ? meta.fragment_mass_tolerance : 13
+    mass_acc_ms1 = meta['precursormasstoleranceunit'].toLowerCase().endsWith('ppm') ? meta['precursormasstolerance'] : 5
```
Note: This fixes a bug where tolerance values were always ignored.

ypriverol commented:
@jspaezp release of `sdrf-pipelines` done with your suggestions https://github.com/bigbio/sdrf-pipelines/releases/tag/v0.0.23

ypriverol commented:
@jspaezp Do you need anything major from us here?

benpullman commented on bin/diann_convert.py:
```diff
@@ -440,46 +597,70 @@ def mztab_PRH(report, pg, index_ref, database, fasta_df):
+    score_looker = ModScoreLooker(report)
     out_mztab_PRH[["modifiedSequence", "best_search_engine_score[1]"]] = out_mztab_PRH.apply(
-        lambda x: PRH_match_report(report, x["accession"]), axis=1, result_type="expand"
+        lambda x: score_looker.get_score(x["accession"]), axis=1, result_type="expand"
```
I think this should use `x["Protein.Ids"]` and not `x["accession"]`

jspaezp replied:
I was wondering whether this was the intended behavior, I kept it as it was on the previous version (adding the change in the next commit: 0b11023)

benpullman commented on bin/diann_convert.py:
```diff
@@ -440,46 +597,70 @@ def mztab_PRH(report, pg, index_ref, database, fasta_df):
 
+    logger.debug("Matching PRH to best search engine score...")
+    score_looker = ModScoreLooker(report)
     out_mztab_PRH[["modifiedSequence", "best_search_engine_score[1]"]] = out_mztab_PRH.apply(
```
I believe `"modifiedSequence"`, `"best_search_engine_score[1]"` are switched, the score should come first based on the output of `score_looker.get_score`

benpullman reviewed (commented):
I have two minor comments around the section that matches PRH to `"best_search_engine_score"`

ypriverol commented:
Hi @jspaezp about the test: 

I was looking for a dataset that allows us to test the pipeline: 
- Most of the datasets in ProteomeXchange has the extension `.d.zip` which means that we should support also that format apart from `.d.gz`

- I recommend as we did with the other DIA experiment creating an SDRF with only one file for the following dataset https://www.ebi.ac.uk/pride/archive/projects/PXD037164 and testing the pipeline with that SDRF.

jspaezp commented:
> * Most of the datasets in ProteomeXchange has the extension `.d.zip` which means that we should support also that format apart from `.d.gz`

that sounds good! the only issue here is that the current base docker image I am using for it does not have zip built-in. solutions: **A**: build and host a new image with zip, **B** add to the documentation the note saying that if `zip` is desired, the container needs to be over-written (pass the burden of zip to the user). 

This would be pretty much the dockerfile for reference.
```
FROM continuumio/miniconda3:23.5.2-0-alpine
RUN apk add --update zip
```


> * I recommend as we did with the other DIA experiment creating an SDRF with only one file for the following dataset https://www.ebi.ac.uk/pride/archive/projects/PXD037164 and testing the pipeline with that SDRF.

This is great, thanks a lot!

jspaezp commented:
I am happy to let you know that I think this is ready to have a final review/merge. 

Since the last update I:
1. Added support for .zip files, I still do not like them and they do not offer any significant compression advantage, therefore PRIDE should suggest people to use an archival-uncompressed method.
2. Added correct handling of file paths when decompressing the files.
3. Made sure pmultiqc worked (thank you for the heavy lifting on that end!)

I am testing it using the following data:

Input sdrf:

```
source name	characteristics[organism]	characteristics[organism part]	characteristics[cell type]	characteristics[disease]	characteristics[cell line]	characteristics[biological replicate]	assay name	comment[technical replicate]	comment[data file]	comment[fraction identifier]	comment[label]	comment[cleavage agent details]	comment[instrument]	comment[proteomics data acquisition method]	comment[modification parameters]	comment[modification parameters]	comment[precursor mass tolerance]	comment[fragment mass tolerance]	comment[file uri]	factor value[phenotype]
Sample 1	homo sapiens	cancer	cells	not applicable	not applicable	1	Run 1	1	3817_TIMS2_2col-80m_37_1_Slot1-46_1_4768.d.zip	1	AC=MS:1002038;NT=label free sample	AC=MS:1001251;NT=Trypsin	AC=MS:1003231;NT=TimsTOF SCP	NT=Data-Independent Acquisition;AC=NCIT:C161786	NT=Oxidation; MT=Variable; TA=M; AC=Unimod:35	NT=Carbamidomethyl; MT=Fixed; TA=C; AC=Unimod:4	15 ppm	15 ppm	https://ftp.pride.ebi.ac.uk/pride/data/archive/2023/05/PXD037164/3817_TIMS2_2col-80m_37_1_Slot1-46_1_4768.d.zip	A
Sample 2	homo sapiens	cancer	cells	not applicable	not applicable	2	Run 2	1	3817_TIMS2_2col-80m_38_2_Slot1-47_1_4816.d.zip	1	AC=MS:1002038;NT=label free sample	AC=MS:1001251;NT=Trypsin	AC=MS:1003231;NT=TimsTOF SCP	NT=Data-Independent Acquisition;AC=NCIT:C161786	NT=Oxidation; MT=Variable; TA=M; AC=Unimod:35	NT=Carbamidomethyl; MT=Fixed; TA=C; AC=Unimod:4	15 ppm	15 ppm	https://ftp.pride.ebi.ac.uk/pride/data/archive/2023/05/PXD037164/3817_TIMS2_2col-80m_38_2_Slot1-47_1_4816.d.zip	A
Sample 3	homo sapiens	cancer	cells	not applicable	not applicable	1	Run 1	1	3817_TIMS2_2col-80m_13_1_Slot1-22_1_4772.d.zip	1	AC=MS:1002038;NT=label free sample	AC=MS:1001251;NT=Trypsin	AC=MS:1003231;NT=TimsTOF SCP	NT=Data-Independent Acquisition;AC=NCIT:C161786	NT=Oxidation; MT=Variable; TA=M; AC=Unimod:35	NT=Carbamidomethyl; MT=Fixed; TA=C; AC=Unimod:4	15 ppm	15 ppm	https://ftp.pride.ebi.ac.uk/pride/data/archive/2023/05/PXD037164/3817_TIMS2_2col-80m_13_1_Slot1-22_1_4772.d.zip	B
Sample 4	homo sapiens	cancer	cells	not applicable	not applicable	2	Run 2	1	3817_TIMS2_2col-80m_14_1_Slot1-23_1_4690.d.zip	1	AC=MS:1002038;NT=label free sample	AC=MS:1001251;NT=Trypsin	AC=MS:1003231;NT=TimsTOF SCP	NT=Data-Independent Acquisition;AC=NCIT:C161786	NT=Oxidation; MT=Variable; TA=M; AC=Unimod:35	NT=Carbamidomethyl; MT=Fixed; TA=C; AC=Unimod:4	15 ppm	15 ppm	https://ftp.pride.ebi.ac.uk/pride/data/archive/2023/05/PXD037164/3817_TIMS2_2col-80m_14_1_Slot1-23_1_4690.d.zip
```

Nextflow config

```
// Pipeline Parameters
params {
    // Input options
    input = 'THAT_SDRF.SDRF.TSV'
    database = 'SOMEHUMANFASTA.fasta'

    skip_post_msstats = true
    add_decoys = true
    acquisition_method = 'dia'

    // DIA-NN related
    mass_acc_automatic = false
    diann_normalize = false
    diann_speclib = ''

}
```

jpfeuffer commented on modules/local/diannsummary/main.nf:
```diff
@@ -27,17 +31,24 @@ process DIANNSUMMARY {
+    # Notes: if .quant files are passed, mzml/.d files are not accessed, so the name needs to be passed but files
+    # do not need to pe present.
+
+    # end, remove when done inspecting.
```
Ready to remove?

jspaezp replied:
Good catch, Will remove on the next commit

jspaezp replied:
removed in f1b0bf786784e6f4193f08a4a3432c42d6602544

jpfeuffer commented on modules/local/dotd_to_mqc/main.nf:
```diff
@@ -0,0 +1,65 @@
+    label 'process_single'
+
+    conda "base::python=3.10"
+    container "continuumio/miniconda3:23.5.2-0-alpine"
```
Is there a singularity container with which we could do an if/else like in the other modules/processes?

Also, can we use a bioconda container?
I see that you probably just need python. Maybe we can ask on nf-core what the preferred way is here.

jspaezp replied:
It is my understanding that ... if the profile is set to singuylarity it will generate singularity containers from the docker containers if passed (although I do not use singularity at the moment so I cannot assure it is the case).

Regarding the bioconda container, I am not all that familiar with that part of the ecosystem. let me know if you have any suggestion for it!

ypriverol replied:
This is the way we should use a container with python 3.10: 

```
conda "conda-forge::python=3.10"
    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
        container "https://depot.galaxyproject.org/singularity/python:3.10"
    } else {
        container "quay.io/biocontainers/python:3.10"
    }
```

jpfeuffer commented on modules/local/pmultiqc/main.nf:
```diff
@@ -25,6 +25,16 @@ process PMULTIQC {
+    set -e
+
+    # leaving here to ease debugging
+    ls -lcth *
```
Still needed?

jspaezp replied:
well ... I like having it when things go wrong but I can see how it can be cluttering if not. (lmk if you think we should remove it)

jpfeuffer replied:
Not necessarily. I think you can leave it.

jpfeuffer commented on modules/local/tdf2mzml/main.nf:
```diff
@@ -0,0 +1,58 @@
+    //    'quay.io/biocontainers/thermorawfileparser:1.3.4--ha8f3691_0' }"
+    container 'mfreitas/tdf2mzml:latest' // I don't know which stable tag to use...
+
+    stageInMode {
```
I think this is not needed. This is only there for the weird behaviour of trfp/mono in general.

jspaezp replied:
perfect, I will move away from it and close the conversation in the next commit.

jpfeuffer commented on nextflow.config:
```diff
@@ -156,6 +156,9 @@ params {
     quantify_decoys          = false
 
+    // Bruker data
+    convert_dotd            = false
```
Ah nice, so with this we could still force conversion to mzml if ever needed?

jspaezp replied:
Indeed that is how it would work! the current implementation uses tdf2mzml to generate the mzML from .d files and that gets passed to DIA-NN. I have not really tested that it gives good results (In my experience it doesnt ... ) but it would be a possibility.

jpfeuffer replied:
Interesting! Yes it is perfectly fine how it is, as a last resort/fallback.

Out of curiosity: According to your experience, what goes wrong when using converted mzmls? Is it just runtime/storage or also quality of the results?

jspaezp replied:
Well ... I have not tried in a while, but If my memory is not failing me it was all of the former to different degrees depending on how the mzml was generaged.
1. If the mzml was generated with arguments/sotware/modes that collapsed the mobility dimension ("centroided over it"), results tend to be poor, I believe this happens becuase the noise that can be easily identified as noise by resolving on the mobility dimension gets over-represented, which leads to "virtually poor" scan quality.
2. In the case of software that does not collapse the mobility dimension, the first thing is that files end up being absoluteley massive. Since each "scan" along the mobility dimension ends up being hundreds of scans in the mzml, the resulting file is the equivalent of having an instrument scanning at +1_000 Hz. Which makes any software reading it very slow and disk usage absolutely horrendous.

I have not tried in a while and we have been exploring different ways to have a good intermediate file format ... the search remains :P

jpfeuffer commented on pyproject.toml:
```diff
@@ -8,3 +8,6 @@ target_version = ["py37", "py38", "py39", "py310"]
 multi_line_output = 3
+
+[tool.ruff]
+line-length = 120
```
I like this length too but we should check with nf-core and their linting.

jspaezp replied:
To be completely honest I don't BUT it was consistent with the rest of the project (and i think its was the configured one in sonatype).

(I personally like 88, but I value consistency more than my personal beliefs lol)

jpfeuffer commented on bin/dotd_2_mqc.py:
```diff
@@ -0,0 +1,362 @@
+logger = getLogger(__name__)
+
+SECOND_RESOLUTION = 5
+MQC_YML = """
```
Why is this needed?
What is the difference to the multiqc yml file in the repo?

jspaezp replied:
This adds as part of the qc information on the chromatogram and total ions that is extracted directly from the .d files in a distributed way.

This specific yml "file" is not used unless multiqc is run in a standalone way (and it lets me test it in a much faster way than having it in a sharded way as part of the pipeline)

jpfeuffer replied:
Ok I guess that's fine. Might just be awkward if the two versions get out-of-sync at some point. Not sure how likely that is.

I didn't check, but is this documented? Maybe add a comment about its "standalone usage" to that variable or the function that writes out its contents.

jspaezp replied:
I am adding a couple of comments and more information on the module docstring for the standalone usage (outside of the quantms pipeline).

jspaezp replied:
on f1b0bf786784e6f4193f08a4a3432c42d6602544

ypriverol reviewed (changes requested):
@jspaezp check some of my comments, solutions for which containers must be used to be able to deploy using conda/docker/singularity. 

Let me know what do you think.

jspaezp commented:
Hello @ypriverol, @jpfeuffer, @benpullman and @WangHong007 Thank you very much for the review and suggestions!

I believe I have incorporated all the suggestions and notes. I am happy to take into account any other suggestions you might have and to work together in the future!

Kindest wishes,
Sebastian
